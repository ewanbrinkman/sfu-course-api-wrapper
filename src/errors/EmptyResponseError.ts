export default class EmptyResponseError extends Error {
    constructor(url: string) {
        super(`API returned empty list when requesting: ${url}`);
        this.name = 'EmptyResponseError';
    }
}
