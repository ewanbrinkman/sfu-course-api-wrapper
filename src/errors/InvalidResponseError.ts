export default class InvalidResponseError extends Error {
    constructor(errorCode: number, url: string) {
        super(`API returned ${errorCode} error when requesting: ${url}`);
        this.name = 'InvalidResponseError';
    }
}
