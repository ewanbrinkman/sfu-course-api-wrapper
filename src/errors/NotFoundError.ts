export default class NotFoundError extends Error {
    constructor(search: string) {
        super(`The search for ${search} didn't return anything.`);
        this.name = 'NotFoundError';
    }
}
