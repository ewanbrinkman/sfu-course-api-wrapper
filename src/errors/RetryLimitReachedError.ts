export default class RetryLimitReachedError extends Error {
    constructor(maxAttempts: number, url: string) {
        super(`Retry limit (${maxAttempts}) reached for: ${url}`);
        this.name = 'RetryLimitReachedError';
    }
}
