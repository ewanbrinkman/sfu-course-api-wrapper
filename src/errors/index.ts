export { default as RetryLimitReachedError } from './RetryLimitReachedError';
export { default as InvalidResponseError } from './InvalidResponseError';
export { default as EmptyResponseError } from './EmptyResponseError';
export { default as NotFoundError } from './NotFoundError';
