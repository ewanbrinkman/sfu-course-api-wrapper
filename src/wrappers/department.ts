import { CourseOutlinesTerm, CourseOutlinesYear } from '@api-types';
import { Department } from '@api';
import wrappers from '@wrappers';
import { NotFoundError } from '@errors';

export default async function department(
    department: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<Department> {
    const departments = await wrappers.departments(year, term);

    const foundDepartment = departments.find((dept) => dept.id === department);
    if (foundDepartment === undefined) {
        throw new NotFoundError(department);
    }

    return foundDepartment;
}
