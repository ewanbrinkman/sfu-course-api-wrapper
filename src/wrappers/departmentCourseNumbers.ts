import { requestSFUAcademicCalendarApiCourses } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawDepartmentCourse,
} from '@api-types';
import { EmptyResponseError } from '@errors';

export default async function departmentCourseNumbers(
    department: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<string[]> {
    const response = await requestSFUAcademicCalendarApiCourses(
        year,
        term.toLowerCase() as CourseOutlinesTerm,
        department.toLowerCase(),
    );
    const rawDepartmentCourses: RawDepartmentCourse[] = await response.json();

    if (rawDepartmentCourses.length === 0) {
        throw new EmptyResponseError(response.url);
    }

    return rawDepartmentCourses.map((rawDepartmentCourse) =>
        rawDepartmentCourse.value.toUpperCase(),
    );
}
