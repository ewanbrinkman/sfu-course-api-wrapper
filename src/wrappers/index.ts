import course from './course';
import courseSection from './courseSection';
import departmentCourseNumbers from './departmentCourseNumbers';
import departmentCourses from './departmentCourses';
import departmentNames from './departments';
import department from './department';
import departments from './departments';
import years from './years';
import terms from './terms';

export default {
    course: course,
    courseSection: courseSection,
    departmentCourseNumbers: departmentCourseNumbers,
    departmentCourses: departmentCourses,
    departmentNames: departmentNames,
    department: department,
    departments: departments,
    terms: terms,
    years: years,
};
