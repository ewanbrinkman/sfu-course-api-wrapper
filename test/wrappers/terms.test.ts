import wrappers from '../../src/wrappers';
import { InvalidResponseError } from '../../src/errors';

describe('terms', () => {
    test('request terms for valid year', async () => {
        const terms = await wrappers.terms(2021);
        expect(terms).toMatchSnapshot();
    }, 30000);

    test('request terms for invalid year', async () => {
        await expect(wrappers.terms(1910)).rejects.toThrow(
            InvalidResponseError,
        );
    }, 30000);
});
