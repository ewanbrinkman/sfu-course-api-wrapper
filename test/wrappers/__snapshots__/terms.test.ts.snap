// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`terms request terms for valid year 1`] = `
[
  "fall",
  "spring",
  "summer",
]
`;
